examples/differentially_private_sgd.py

```diff
@@ -134,33 +133,30 @@ def accuracy(params, batch):
   return jnp.mean(predicted_class == target_class)
 
 
+def clipped_grad(params, l2_norm_clip, single_example_batch):
+  """Evaluate gradient for a single-example batch and clip its grad norm."""
+  grads = grad(loss)(params, single_example_batch)
+  nonempty_grads, tree_def = tree_flatten(grads)
+  total_grad_norm = jnp.linalg.norm(
+      [jnp.linalg.norm(neg.ravel()) for neg in nonempty_grads])
+  divisor = jnp.max((total_grad_norm / l2_norm_clip, 1.))
+  normalized_nonempty_grads = [g / divisor for g in nonempty_grads]
+  return tree_unflatten(tree_def, normalized_nonempty_grads)
+
+
 def private_grad(params, batch, rng, l2_norm_clip, noise_multiplier,
                  batch_size):
   """Return differentially private gradients for params, evaluated on batch."""
-
-  def _clipped_grad(params, single_example_batch):
-    """Evaluate gradient for a single-example batch and clip its grad norm."""
-    grads = grad(loss)(params, single_example_batch)
-
-    nonempty_grads, tree_def = tree_util.tree_flatten(grads)
-    total_grad_norm = jnp.linalg.norm(
-        [jnp.linalg.norm(neg.ravel()) for neg in nonempty_grads])
-    divisor = stop_gradient(jnp.amax((total_grad_norm / l2_norm_clip, 1.)))
-    normalized_nonempty_grads = [g / divisor for g in nonempty_grads]
-    return tree_util.tree_unflatten(tree_def, normalized_nonempty_grads)
-
-  px_clipped_grad_fn = vmap(partial(_clipped_grad, params))
-  std_dev = l2_norm_clip * noise_multiplier
-  noise_ = lambda n: n + std_dev * random.normal(rng, n.shape)
```

Comment:
Fixing `rng` in `noise_` will introduce dependencies on the noise added across layers in a single update.
The Gaussian mechanism in differential privacy requires independently sampled noise in each parameter.

Reply:
Indeed! (This is a good diagnosis of the original problem this PR is fixing, since this comment is on deleted code.)